core/src/main/java/com/linecorp/armeria/common/HttpRequest.java

```diff
@@ -519,4 +523,54 @@ default HttpRequestDuplicator toDuplicator(EventExecutor executor, long maxReque
                                         Function<? super HttpData, ? extends ByteBuf> byteBufConverter) {
         return new DecodedHttpStreamMessage<>(this, decoder, alloc, byteBufConverter);
     }
+
+    /**
+     * Transforms the {@link ResponseHeaders} of this {@link HttpRequest} by applying the specified
+     * {@link Function}.
+     */
+    default HttpRequest mapHeaders(Function<? super RequestHeaders, ? extends RequestHeaders> function) {
+        requireNonNull(function, "function");
+        final RequestHeaders transformed = function.apply(headers());
+        requireNonNull(transformed, "function.apply() returned null");
+        return of(transformed, this);
+    }
+
+    /**
+     * Transforms the {@link HttpData}s emitted by this {@link HttpRequest} by applying the specified
+     * {@link Function}.
+     */
+    default HttpRequest mapData(Function<? super HttpData, ? extends HttpData> function) {
+        requireNonNull(function, "function");
+        final StreamMessage<HttpObject> stream =
+                map(obj -> obj instanceof HttpData ? function.apply((HttpData) obj) : obj);
```

Comment:
I think this does not fuse when multiple
`request.mapData(...).mapData(...).mapTrailers(...)` is called which means it creates additional instance every time.
How about checking if the delegate of `StreamMessageBasedHttpRequest` is fuseable and fusing it ?